Material Refresh

t895 opened the pull request:
I updated a few parts of the app to make things a little more modern
- Material 3 theming
- Edge to edge
- Predictive back
- Fixed an issue where the balanced indicator would repeatedly vibrate the device
- Use lighter "tick" haptics

<details>
  <summary>Images</summary>
  
![image](https://github.com/user-attachments/assets/802deca7-660b-4b74-8291-8b20e70687cf)
![image](https://github.com/user-attachments/assets/a7008412-88f1-4a9e-82a9-1d20a2661313)
![image](https://github.com/user-attachments/assets/12440229-0570-4172-99ee-d8536d7a46a3)
![image](https://github.com/user-attachments/assets/32a7c279-d798-4bab-a460-0d38a4fea90c)

</details>

Apologies for the seemingly large PR. It's primarily from reformatting xml layouts which don't change much.

t895 commented:
Sorry I made this a little opinionated, but let me elaborate

- Background color for image viewer
I intended to remove the light gray background simply because it seems odd compared to any other image viewer that I've seen.
Would it be reasonable to leave as-is here?

- More settings button shifting
I can't replicate this behavior on my device. Could you send a video or further explanation?

- Vertical padding in settings dialog
I see what you mean where there's extra padding just below the divider. I'll fix that.

- Gyroscope issues
I'll remove that. Good to know that it's been fixed elsewhere.

- Settings dialog background
I'll change that back

- Tab indicator color
I'm using the primary color right now, but I will try out some other colors to see if I could get closer to what you're thinking

MHShetty commented:
HI @t895,

Thanks for addressing the feedbacks!

> Background color for image viewer
> I intended to remove the light gray background simply because it seems odd compared to any other image viewer that I've seen.
> Would it be reasonable to leave as-is here?

I had initially used a specific gray background so that it would be easy to recognize the differentiate the bounds of the image from the background in case of a dark image with little details. I personally feel it's better to keep it that way as have been used to seeing it that way for a long time. We could consider the opinion of other users before we are able to decide on this

MHShetty commented:
Hi @t895,

Thanks for retaining the original background transition in your latest set of commits! Although the UI looks a bit off right now primarily because the action bar color has been set to transparent (unlike the original translucent color), and there's no to little padding to center the image/thumbnail preview correctly, which is more visible as we view images of different aspect ratios. 

The padding issue could have occurred because of introducing edge to edge, or because the padding that was previously being applied might have been accidentally removed. The latest release correctly centers the image as expected (as a reference for the expected behavior)

Adding the original translucent action bar color should fix the first issue pertaining to the action bar. For the second issue, please make sure the padding is applied on the correct view to ensure that zooming in the preview utilizes the entire screen and the newly applied padding doesn't cause a fixed gap to occur at the top of the preview.

t895 commented:
I'll check those out soon

Also sorry if I gave the impression that I was done with working over your suggestions. I'm just going to be busy until this weekend and then I'll be able to finish things up.

MHShetty commented:
Ah there's no need to apologize at all. We appreciate and value the time and efforts you have been putting in to contributing for GrapheneOS. Thanks a lot for the contributions!

MHShetty commented:
Ah, there’s no need to apologize at all. We greatly appreciate and value the time and effort you have put into contributing to GrapheneOS. Thank you so much for your contributions!

t895 commented:
I just fixed up the weird insets for the camera previews and the bottom tab layout

I still need to fix the issues with the settings dialog. The flags for ignoring system bars are really messy.

I also wanted to get your opinion on this change. I know that you wanted to get a translucent top app bar, but since we're edge-to-edge now, I thought we could get something a little cleaner looking. If you don't like the look, I could just change it to the previous solid color.

https://github.com/user-attachments/assets/27a6a683-e753-487c-9fee-63748a91e7d7

MHShetty commented:
Hi @t895,

The primary reason to go with a translucent action bar was because it looks off with the gray background. The image preview in the video seems to be spanning the entire screen causing the background to remain hidden at the moment. Ideally the gray background would be visible at the default/minimum zoom level making it look a bit off. I feel we'll be better off going with gray background and a translucent bar as before

t895 commented:
I'm unsure about what you'd like from the indicator color, so I put together a few combinations for you to browse. This shows a pretty standard blue theme and what I would consider to be "worst case scenario" as a monochrome theme.

Let me know what you think

<details>

<summary>Catalog</summary>

## Primary

### Monochrome
![image](https://github.com/user-attachments/assets/36e95f50-531e-4ed1-982d-b76e2c98fdaa)

### Blue
![image](https://github.com/user-attachments/assets/f622dc02-1ec4-4a90-9ea8-55072d8baf20)

## PrimaryFixed

### Monochrome
![image](https://github.com/user-attachments/assets/51546fb9-5faf-4cdd-972b-d0372ef96ea8)

### Blue
![image](https://github.com/user-attachments/assets/c2f1cf28-df3d-4a93-bf29-2475a6ece7ed)

## Secondary

### Monochrome
![image](https://github.com/user-attachments/assets/ca0b76a3-85d3-4072-bfc0-6456f8d2d3d2)

### Blue
![image](https://github.com/user-attachments/assets/81665114-f33c-49d9-8aa5-9169665a667d)

## Secondary Fixed

### Monochrome
![image](https://github.com/user-attachments/assets/554a3f4a-c0b1-492c-ae00-f6462261a4d9)

### Blue
![image](https://github.com/user-attachments/assets/915063b7-dab5-496e-88c5-9cf0d58efbaa)

## Tertiary

### Monochrome
![image](https://github.com/user-attachments/assets/9fe2d9ae-6f08-4b94-b41f-f2306ab09536)

### Blue
![image](https://github.com/user-attachments/assets/c1e848ab-0eba-41ae-8142-06bc10a26bb9)

## TertiaryFixed

### Monochrome
![image](https://github.com/user-attachments/assets/a7aedc78-d9c5-40b6-9453-44464840fc5e)

### Blue
![image](https://github.com/user-attachments/assets/49bcfe62-3fab-40b5-81a4-c4dfcba31a5e)

</details>

t895 commented:
Everything except for the indicator has been addressed now. I've also updated the images in the opening comment to match as well.

Let me know if you spot any other oddities.

MHShetty commented:
* 

* The preview seems to have slid down unlike how it was aligned to the top as before. Please keep the alignment the same as before.

* The actions of the action bar in the in-app gallery seem too close to the bottom end, adding sufficient vertical padding should help resolve this issue.

* The video player too seems to have a black background unlike the gray one which was previously followed. Please update the background color for that too.

Also, please refrain from making new changes to the PR now. With each iteration, it is primarily introducing new regressions, causing more harm than help.

If possible, only keep the changes related to Material Theming of the app in this PR and separate off the extra changes as a different set of PRs. This would make testing and accepting changes a lot easier for us

t895 commented:
Addressed your feedback in the latest push

Also, I wasn't trying to add new features to this PR. I was trying to solve an issue with edge-to-edge in a way that you ended up disliking. I've included a more involved solution for making sure that we lay out all of the components of the main activity correctly. The app looks the same but if you use it between devices with a notch and ones without, `previewView` will properly take up as much of the screen as possible while avoiding the notch.

As for reviewability, I'm hesitant to break this into a material and an edge-to-edge PR. (Although I'm fine with breaking off the other changes)
The issue being that the original views that were being used do not work with edge-to-edge and it would result in doing extra work for no good reason. Would you instead be ok if I broke this up per-activity?

MHShetty commented:
> As for reviewability, I'm hesitant to break this into material and edge-to-edge PRs. (Although I'm fine with breaking off the other changes)

I was referring to the other changes (such as the main overlay getting immediately applied, the changes that caused the preview to be repositioned/cover the entire screen, etc.)

We need to support edge to edge as devices starting from Android 15 have enforced it, therefore we want the designs to be dealing with that.

> The issue being that the original views that were being used do not work with edge-to-edge and splitting things up would result in doing extra work for no good reason. Would you instead be ok if I broke this up per-activity?

The UI of main activity has been covering most of the screen already, so apart from setting the top insets correctly it shouldn't be much of an issue otherwise, unless you're changing the UI to address a different issue altogether, causing more regressions to deal with each iteration which could be avoided.

It's not about disliking things. With every change you are making we need to make sure that there aren't any other regressions as all the components of the app are visually/functionally dependent on each other, causing the entire process of accepting an PR to take longer than otherwise. Once the main issue/feature of the PR is addressed, we can always deal with new things you might wish to add in the app in a separate pull request

t895 commented:
> I was referring to the other changes (such as the main overlay getting immediately applied, the changes that caused the preview to be repositioned/cover the entire screen, etc.)
> We need to support edge to edge as devices starting from Android 15 have enforced it, therefore we want the designs to be dealing with that.

I think I have miscommunicated things here. The main overlay commit is to address an issue with the previous edge to edge commit that I felt needed further explanation.

> The UI of main activity has been covering most of the screen already, so apart from setting the top insets correctly it shouldn't be much of an issue otherwise, unless you're changing the UI to address a different issue altogether, causing more regressions to deal with each iteration which could be avoided.

Making the activity edge to edge causes issues that I have to address with more involved solutions than just adding insets. I need to make sure that we avoid the notch whenever possible and that the tab layout and shutter button are put in the same position as the release version. Otherwise, I would be introducing worse regressions. Sorry that I didn't spot them before.

> It's not about disliking things. With every change you are making we need to make sure that there aren't any other regressions as all the components of the app are visually/functionally dependent on each other, causing the entire process of accepting an PR to take longer than otherwise. Once the main issue/feature of the PR is addressed, we can always deal with new things you might wish to add in the app in a separate pull request

The screenshots that I show don't communicate this, but there is a notch at the top of the screen on my Pixel 7. We need to avoid it unless there isn't the screen space to allow for it. I tried introducing a commit that would fix this by biasing the camera preview away from the notch but that was poor. So, I went back and completed a more involved solution that avoids the notch exactly. On screens that don't have a cutout, it goes right to the top.

MHShetty commented:
![A screenshot](https://github.com/user-attachments/assets/e92ca610-a3db-4e85-821c-fa229ab2d580)

The drop-down seems to be too close to the bottom edge of the top insets/black area. Making it appear centered as before would be better. The zoom and exposure panel seem to be going below the preview view's height.

MHShetty commented:
> Making the activity edge to edge causes issues that I have to address with more involved solutions than just adding insets. I need to make sure that we avoid the notch whenever possible and that the tab layout and shutter button are put in the same position as the release version

Are you facing any issues apart from correctly calculating the height of the top insets in the MainActivity?

t895 commented:
Wait, I see what you've been trying to point out now. It looks like I've lost the plot a bit with the extra changes I've made in the main activity. I'm gonna revert most of that and get things back on track.

t895 commented:
Ok, that makes much more sense now.
You were right and I was going overboard with changes. How does this look now?

MHShetty commented:
Yes, this is a lot better than before. Thanks for making the changes!

MHShetty commented:
Hey @t895,

https://github.com/GrapheneOS/Camera/pull/480/commits/6ca603d40819211949e2a33e9f3f3f7fdb276697

Not really sure why you have created multiple drawable files for the different colors but the same icon otherwise. There is a long list of unexplained changes in this commit. How many of these changes are actually necessary for material theming? Please only keep the changes that is needed for material theming and make a separate PR for the other changes. If there are none, then please shift and split the commit in a different PR, grouping similar changes in a single commit which would make your changes simpler to understand

t895 commented:
Oh, let me explain here and I'll leave a comment on that commit as well.

They need to be duplicated because there are some icons that need to both dynamically change and always be white. For example, since the torch icon was reused between the QR scanner and the icon shown in the settings dialog, they needed to be split because the QR scanner's icon must always be white and the settings dialog's icon changes with the system colors. There is no way to change the color of an icon in xml layouts with <item> tags so I just dealt with it this way.

I also removed the while color that you had in `colors.xml` since `@android:colors/white` already exists and the usage of `Theme_Dialog` because that was no longer in use.

However, the layout changes were from running the built-in xml formatter for Android Studio. I'll go back quick and keep it to the bare minimum because I see that it could make things annoying to review.

MHShetty commented:
Apart from that the code changes seem mostly fine now. 

Mostly the last set of reviews/feedback (related to visual appearance based on the changes made in this PR earlier):

* The padding with the new options still seems to be inconsistent across different types of options however.

![A screenshot](https://github.com/user-attachments/assets/172c3e43-513d-4dbd-8fec-b2bb5e7b9e74)
![A screenshot](https://github.com/user-attachments/assets/03735e0f-ffa8-414e-b8e0-f46ef6f05f11)

* The toggle buttons seem to be a bit unusually larger than the text beside them in the settings dialog. Maybe making the toggle a bit smaller or the text a bit larger could help make it look better.

* The image seems to be not so vertically centered in the in-app gallery, possibly being caused due to the extra height edge-to-edge has introduced. Whatever changes are made to fix this, please make sure it doesn't affect how the zooming is expected to work.

MHShetty commented:
> They need to be duplicated because there are some icons that need to both dynamically change and always be white. For example, since the torch icon was reused between the QR scanner and the icon shown in the settings dialog, they needed to be split because the QR scanner's icon must always be white and the settings dialog's icon changes with the system colors. There is no way to change the color of an icon in xml layouts with tags so I just dealt with it this way.

Ohh that makes sense now

> However, the layout changes were from running the built-in xml formatter for Android Studio. I'll go back quick and keep it to the bare minimum because I see that it could make things annoying to review.

Sure, please make sure you're grouping/splitting changes that are similar and only keeping the ones that are necessary for this pull request

t895 commented:
> * The padding with the new options still seems to be inconsistent across different types of options, however.
>   (The spacing of the dropdown seems to be a bit lesser than the others in the first image, the other image seems to be inconsistent in general)
> * The toggle buttons seem to be a bit unusually larger than the text beside them in the settings dialog. Maybe making the toggle a bit smaller or the text a bit larger could help make it look better.

These two are the same problem because the material switch is pretty big. I also can't shrink the new components to be small enough to appear even. Would it be reasonable to (while keeping the new components as small as possible) make each item take up the same amount of space?

> * The image seems to be not so vertically centered in the in-app gallery, possibly being caused due to the extra height edge-to-edge has introduced. Whatever changes are made to fix this, please make sure it doesn't affect how the zooming is expected to work.

As far as I can tell, there are no problems with zooming with this change. You'd know better than me.

MHShetty commented:
> These two are the same problem because the material switch and the material radio button are pretty big. I also can't shrink the new components to be small enough to appear even. Would it be reasonable to (while keeping the new components as small as possible) make each item take up the same amount of space?

Yes, both of these problems are being caused due to replacing existing components with their material counterparts. However, fixing the padding/height issue won't fix the text/toggle size issue, making them two different issues to work on.

> As far as I can tell, there are no problems with zooming with this change

I'm referring to the centering of the image in the in-app gallery. The image currently looks more aligned towards the top than just roughly being at the center. While applying a fix to the issue that I have described in the previous sentence, please make sure it doesn't cause a fixed inset to exist on the top while zooming.

t895 commented:
Just a quick update

I'm still working on the image viewer padding. Just running into some weird behavior where the offset we're seeing is only happening when using gesture navigation. I should have a fix later today.

Otherwise, I've fixed up the padding consistency issue for the settings dialog. I also caught and fixed a regression with the QR result dialog.

t895 commented:
Ok I feel kinda dumb for not noticing this, but the padding on the image view actually isn't a problem.
In the release version, the image view receives bottom padding based on the height of the navigation bar.
In this PR, the image view is now consistently in the center of the screen.

MHShetty commented:
Hi @t895,

Not really sure if you're done with the changes, but the latest changes might have caused the "More Settings" button to not appear on certain screen sizes when the device gets rotated 90 degrees to make it horizontal.

The padding/height of each item seems to be a lot more similar now than before. Great work!

t895 commented:
I'm done with changes but I'll fix that more settings button asap

t895 commented:
It's fixed up now. Lmk what you think.

MHShetty commented:
Hi @t895,

Thanks for making those changes. The PR now seems to be mostly done, but just caught one issue with the edge to edge implementation:

![Screenshot_20241119-033600](https://github.com/user-attachments/assets/6c1868bf-5295-4b4a-8abd-2572f1ebf328)

The edge to edge seems to be partly broken on smaller screen devices. The bottom tab layout isn't really accessible here. It might be safer to keep the lower part padded for the insets in the main activity to help avoid the users face such issues.

![Screenshot_20241119-033717](https://github.com/user-attachments/assets/a848228d-abdd-42fd-b29d-67d0832b1b3f)

Please make sure the settings drop-down is correctly centered in the top gap with minimal/sufficient vertical margin

t895 commented:
Oh weird

I'll test it in an emulator

t895 commented:
I fixed the issue with the tab layout, but the settings drop down button has always been in that position on small devices.

MHShetty commented:
The bottom UI gets fixed to the bottom of the activity. It doesn't always get placed that way, but it happens a lot of the times with the latest changes

![Screenshot_20241119-154133.png](https://github.com/user-attachments/assets/55397925-d2dd-4cdb-a34f-11507d2ba7e1)

Navigating between different activities and coming back to the main activity is one of the ways to reproduce it

MHShetty commented:
Btw, thanks a lot for preparing and sharing the color catalog here https://github.com/GrapheneOS/Camera/pull/480#issuecomment-2465817436

For the primary color not matching our existing shade, I don't think there's any reliable solution that works well across all colors, it might best to continue with the dynamic primary color given by Android. Although, we could let the user choose between using dynamic color and reverting to the old theme by conditionally applying them instead while keeping the fallback to old colors that were being originally used

MHShetty commented:
Btw, thanks a lot for preparing and sharing the color catalog here in https://github.com/GrapheneOS/Camera/pull/480#issuecomment-2465817436

For the issue of the dynamic primary color not matching our existing shade, I don't think there's any reliable solution that works well across all colors, it might be best to continue with the dynamic primary color given by Android. 

Although, we could let the user choose between using dynamic color and reverting to the old theme by conditionally applying them instead while keeping the fallback to old colors that were being originally used

MHShetty commented on app/src/main/java/app/grapheneos/camera/ui/activities/InAppGallery.kt:
```diff
@@ -404,10 +432,17 @@ class InAppGallery : AppCompatActivity() {
         startActivity(Intent.createChooser(share, getString(R.string.share_image)))
     }
 
+    @SuppressLint("WrongConstant")
```
Unable to really see its usage but I think SuppressLint("WrongConstant") could be removed from here

t895 commented:
> It would be better to have it placed as before with the required bottom padding for the bottom insets

I reverted it back to the original

t895 commented on app/src/main/java/app/grapheneos/camera/ui/activities/VideoPlayer.kt:
```diff
@@ -26,8 +34,21 @@ class VideoPlayer : AppCompatActivity() {
+        // This is a deprecated constant but Google uses it themselves in Photos and results in
+        // much nicer UX. Suppress the warning until we get a decent API.
+        windowInsetsController.systemBarsBehavior =
+            WindowInsetsControllerCompat.BEHAVIOR_SHOW_BARS_BY_TOUCH
```
Oh sorry, I meant to revert this earlier. It's fixed now.

MHShetty commented:
> I reverted it back to the original

Oh as in the version before the bottom insets issue was addressed?

MHShetty commented:
Could adding bottom padding to the main container and keep the original changes help resolve the issue without adding new regressions?

t895 commented:
Sorry, in trying to respond to you quickly, I missed something.
I made that original change because it would remove the black navigation bar that you would see if you used the front facing camera and enabled self illumination. Give me a moment to go back and address this properly.

MHShetty commented:
Yes, the solution resolves the final issue with the edge-to-edge implementation. I think the PR is ready to be merged.

Thank you for all your time, patience and efforts. Appreciate and value it all!

MHShetty commented:
You're always welcome, glad to have your kind words, support and valuable contributions too mate!

muhomorr commented:
@t895 "Copy to clipboard" button in ImageCapturer error dialog no longer works after these changes. Tapping on it simply dismisses the dialog. Interestingly, tapping above that button does activate it.

muhomorr commented:
Huge switches for "Include Audio" and "Enable EIS" items in video quick settings menu look wrong. Their large size also makes the "Focus Timeout" setting invisible, it's easy to miss that it even exists. The only hint of that it's there is a faint scrollbar on the right of quick settings popup that quickly disappears.

t895 commented:
> "Copy to clipboard" button in ImageCapturer error dialog no longer works after these changes. Tapping on it simply dismisses the dialog. Interestingly, tapping above that button does activate it.

I couldn't get a camera error to happen but I tried just running the dialog in the main activity and I couldn't get the issue to happen on my devices. Could you explain a bit more about what you did to trigger that? I know that these dialogs can be really finicky with shortEdges so I don't doubt that it happened.

> Switching between camera modes looks wrong now sometimes because mode name (CAMERA/VIDEO/etc) color changes too late or too early.

I don't believe this is a problem? I didn't make a change to the timing of the text color animation when switching between tabs. This is an issue with the tab layout that we can't change.
The problem was just invisible before because the text color was always the same, but you can see the same issue if you use the front facing camera with self illumination in the release build.

> Huge switches for "Include Audio" and "Enable EIS" items in video quick settings menu look wrong. Their large size also makes the "Focus Timeout" setting invisible, it's easy to miss that it even exists. The only hint of that it's there is a faint scrollbar on the right of quick settings popup that quickly disappears.

I pushed a commit to try and address this. I shrunk the switch and always show the scrollbar when the list is scrollable now.

MHShetty commented:
> Could you explain a bit more about what you did to trigger that?  

One of the ways to trigger it is by switching modes while clicking a photo (easier to reproduce in night mode)

t895 commented:
I went with a more involved fix since this can potentially effect any material alert dialog that appears in the camera preview screen

muhomorr commented:
> I don't believe this is a problem? I didn't make a change to the timing of the text color animation when switching between tabs. This is an issue with the tab layout that we can't change.
The problem was just invisible before because the text color was always the same, but you can see the same issue if you use the front facing camera with self illumination in the release build.

It's true that the text color animation is present in the release build too, but it's noticeable there only in one mode that is disabled by default. These changes make this issue noticeable in all modes.

> I pushed a commit to try and address this. I shrunk the switch and always show the scrollbar when the list is scrollable now.

Maybe it's better to not use a switch at all? Pixel Camera doesn't use them in its roughly equivalent UI element.

t895 commented:
> It's true that the text color animation issue is present in the release build too, but it's noticeable there only in one mode that is disabled by default. These changes make this issue noticeable in all modes.

Given that this is an issue with the component itself, I could just revert this color change. Making an entirely new component is way out of the scope of this PR.

> Maybe it's better to not use a switch at all? Pixel Camera doesn't use them in its roughly equivalent UI element.

I feel like we could argue similarly about the right component to use for every part of the settings dialog since it has pretty tight size constraints. If we went with what the Pixel Camera has, we would run into the same problem.
I think this is a reasonable in-between before larger changes could be made.

t895 commented:
Personally, I really don't like working with Android Views, but when Compose is introduced into this project with #468, I'd be interested in rewriting the camera UI with Compose.

t895 commented:
Well if y'all need help with Compose or anything else with your apps, ping me and I'd be happy to lend a hand

MHShetty commented:
Hi @t895,

Thanks a lot for showing interest in contributing to GrapheneOS, appreciate and value it all. We are already working on porting our camera app to Jetpack Compose, will surely let you know if there's any help needed for testing the app or anything related
